lib/vector-core/src/schema/definition.rs

```diff
@@ -1,9 +1,326 @@
-#[derive(Debug, Clone, PartialEq)]
-pub struct Definition;
+use std::collections::{HashMap, HashSet};
+
+use lookup::LookupBuf;
+use value::{
+    kind::{merge, nest, Collection, Field, Unknown},
+    Kind,
+};
+
+/// The definition of a schema.
+///
+/// This struct contains all the information needed to inspect the schema of an event emitted by
+/// a source/transform.
+#[derive(Clone, Debug, PartialEq)]
+pub struct Definition {
+    /// The collection of fields and their types stored in the event.
+    collection: Collection<Field>,
+
+    /// Semantic meaning assigned to fields within the collection.
+    ///
+    /// The value within this map points to a path inside the `collection`. It is an invalid state
+    /// for there to be a meaning pointing to a non-existing path in the collection.
+    meaning: HashMap<String, LookupBuf>,
+
+    /// A list of paths that are allowed to be missing.
+    ///
+    /// The key in this set points to a path inside the `collection`. It is an invalid state for
+    /// there to be a key pointing to a non-existing path in the collection.
+    optional: HashSet<LookupBuf>,
```

Comment:
I might just be missing somethign simple, but why is this a separate field? I would have expected this to be extra information stored in `collection`, something like `collection` storing `known: BTreeMap<T, (Kind, bool)>`.

Reply:
That would only allow for the first level in a collection to be optional.

Reply:
Ah, sorry, I now see that you're talking about the type `Collection`, not the field above named `collection`.

`Collection` has no concept of "optional" because it's not meant to be used for that purpose. It's part of the "value type" API, which has no use of optional fields. It _does_ have the concept of a field that is `null`, but that is separate from a field being optional.

It could, of course, embed that information, and I considered it, but I found that a leaky abstraction implemented in the wrong place. It is true that the result of that conclusion is that this type (`Definition`) has to be strictly vetted not to invalidate the invariant that any paths in this set have to be present in the collection, but that was a trade-off I thought was worth making.